core/src/main/scala/kafka/server/KafkaServer.scala

```diff
@@ -254,7 +257,8 @@ class KafkaServer(val config: KafkaConfig, time: Time = Time.SYSTEM, threadNameP
         replicaManager.startup()
 
         val brokerInfo = createBrokerInfo
-        zkClient.registerBroker(brokerInfo)
+        val epoch = zkClient.registerBroker(brokerInfo)
+        brokerEpoch.update(epoch)
```

Comment:
Currently `brokerEpoch` can be updated in `KafkaServer.startup()` as well as `RegisterBrokerAndReelect` in the controller, and it can be read by `ReplicaManager` in order to reject outdated control requests. If we store `brokerEpoch` in `KafkaController`, it requires passing the `KafkaController` object to `ReplicaManager` just for reading `brokerEpoch`. I think use a helper class is simpler in this case. What do you think?

Reply:
Miss your next comment. If we do the check in `KafkaApis`, then you are right. Please ignore the comment I just wrote down.

Reply:
Fixed.